libbeat/publisher/pipeline/processor.go

```diff
@@ -290,6 +295,23 @@ func makeAddDynMetaProcessor(
 	})
 }
 
+func makeAddAgentMetadataProcessor(info beat.Info) *processorFn {
+	metadata := common.MapStr{
+		"type":         info.Beat,
+		"ephemeral_id": info.EphemeralID.String(),
+		"hostname":     info.Hostname,
+		"id":           info.ID.String(),
+		"version":      info.Version,
+	}
+	if info.Name != info.Hostname {
+		metadata.Put("name", info.Name)
+	}
+	return newProcessor("add_agent_metadata", func(event *beat.Event) (*beat.Event, error) {
+		_, err := event.Fields.Put("agent", metadata)
```

Comment:
Use `metadata.Clone()` or pass the `needsCopy` bool to make sure you don't need to clone. If user adds processors adding/removing fields from the `agent` namespace beats will panic here or in the json encoder due to concurrent access.

Why use put instead of `(Deep)Update`? This is another change in semantics. With put you overwrite `agent`, while update (old behavior) would merge the contents if metadata if `agent` is already known to `Fields`.

Reply:
Good catch on both, will make those changes.

Reply:
done